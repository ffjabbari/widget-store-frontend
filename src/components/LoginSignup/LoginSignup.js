import { useCallback } from 'react';
import { Tabs, Tab, Row, Col, Container, Modal } from 'react-bootstrap';
import { useHistory } from 'react-router-dom';

// local imports
import Login from './Login/Login';
import Signup from './Signup/Signup';

const LoginSignup = () => {
  const history = useHistory();

  // close modal function
  const handleClose = useCallback(
    e => {
      return history.goBack();
    },
    [history]
  );

  return (
    <Container>
      <Row>
        <Col sm="12" md={{ size: 6, offset: 3 }} className="rounded">
          <div>
            <Modal show={true} onHide={handleClose}>
              <Modal.Body>
                <Tabs defaultActiveKey="login">
                  <Tab eventKey="login" title="Login" className="p-5">
                    <Login handleClose={handleClose} />
                  </Tab>
                  <Tab eventKey="signup" title="Signup" className="p-5">
                    <Signup handleClose={handleClose} />
                  </Tab>
                </Tabs>
              </Modal.Body>
            </Modal>
          </div>
        </Col>
      </Row>
    </Container>
  );
};

export default LoginSignup;
