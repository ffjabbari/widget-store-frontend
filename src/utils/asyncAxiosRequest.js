/** Asynchronous axios request function. */

import axios from 'axios';

const BASE_URL = process.env.REACT_APP_BASE_URL || 'http://localhost:3001';

/** asyncAxiosRequest()
 *
 * Function to abstract error handling of axios requests.
 *
 * @param {string} url
 * @param {string} method
 * @param {object} data
 * @param {object} params
 * @param {string} baseURL
 */

export const asyncAxiosRequest = async (
  url,
  method,
  data,
  params,
  baseURL = BASE_URL
) => {
  try {
    const resp = await axios({ url, method, data, params, baseURL });
    // console.log(resp);
    return resp;
  } catch (error) {
    return { error };
  }
};
