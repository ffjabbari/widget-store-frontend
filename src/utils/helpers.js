import { Redirect, Link } from 'react-router-dom';
import { motion } from 'framer-motion';

/**
 * getPathRoot()
 *
 * Given a path return the "root" argument.
 * For example the "root" of /shop/product/206 would be "/shop".
 *
 * @param {string} path
 */

export const getPathRoot = path => {
  return path.match(/^(\/\w*)\/?.*$/)[1];
};

/**
 * <FramerRedirect to="" />
 *
 * Component to force react-router-dom redirect when framer-motion
 * prevents a redirect from executing.
 *
 * If regular <Redirect /> fails in this environment use this redirect instead.
 *
 * @param {any} props
 *
 * returns: motion.div wrapped redirect component
 */
export const FramerRedirect = props => (
  <motion.div exit="undefined">
    <Redirect {...props} />
  </motion.div>
);
export const FramerLink = props => (
  <motion.div exit="undefined">
    <Link {...props} />
  </motion.div>
);

/**
 * animateVariant()
 *
 * Animation function to turn on an animation state then turn it back
 * off after a delay. The state will be set "active" then back to "default".
 *
 * If inital state provided by setterFunct is false this function will skip
 * setting the animation state "active" instead set it to "default".
 *
 * If a callback is provided it will be called with any arguments
 * after setting state back to "default".
 *
 *
 * @param {function} setterFunct - useState setter funct to set animation state
 * @param {integer} delay - delay of setterFunct being called
 * @param {funct} callback - function to call upon completion
 * @param {array} args - array of arguments for callback function
 */

export const animateVariant = (setterFunct, delay, callback, args = []) => {
  // if animation state is false set to 'default' else 'active'
  setterFunct(state => (state === false ? 'default' : 'active'));
  // after delay set back to "default"
  return setTimeout(() => {
    setterFunct('default');
    if (callback) callback(...args);
  }, delay);
};
