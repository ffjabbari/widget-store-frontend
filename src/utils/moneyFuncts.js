/** Functions for basic money calculations for widget-store. */

import Decimal from 'decimal.js';

/* SF sales tax. */
const SF_SALES_TAX = '0.085';

/** calculateDiscountPrice()
 *
 * Calculate discounted price from product object.
 * Product object should have properties "price"
 * and "discount".
 * @param {object} product
 *
 * returns: string
 */

export function calculateDiscountPrice(product) {
  if (product.discount === '0.00') return product.price;
  const price = new Decimal(product.price);
  return price.minus(price.times(product.discount)).toFixed(2);
}

/** calculateSubtotal()
 *
 * Calculate items subtotal from cart.items object.
 * Each cart item should have properties "price", "discount",
 * and "quantity" of items ordered.
 *
 * @param {object} items {product_id: item, product_id: item}
 *
 * returns: string
 */

export function calculateSubtotal(items) {
  return Object.values(items)
    .reduce((acc, product) => {
      const price = new Decimal(calculateDiscountPrice(product));
      const itemTotal = price.times(product.quantity);
      return itemTotal.plus(acc);
    }, 0)
    .toFixed(2);
}

/** calculateDiscountPercent()
 *
 * Calculate discount as an integer percent from
 * product object.discount.
 *
 * E.g. 0.15 --> 15
 *
 * returns: string
 */

export function calculateDiscountPercent(product) {
  const discount = new Decimal(product.discount);
  return discount.times(100).toFixed(0);
}

/** getDiscountData()
 *
 * Return discounted price and discount as a percent
 * from product obj.
 * @param {object} product
 *
 * returns { discountPrice, percent }
 */

export function getDiscountData(product) {
  return {
    discountPrice: calculateDiscountPrice(product),
    percent: calculateDiscountPercent(product),
  };
}

/** calculateTax()
 *
 * Calculate tax if customer is in California.
 * @param {string} subtotal
 * @param {string} state
 *
 * returns: string
 */

export function calculateTax(subtotal, state) {
  if (!/^ca(lifornia)?$/i.test(state)) return '0.00';
  subtotal = new Decimal(subtotal);
  return subtotal.times(SF_SALES_TAX).toFixed(2);
}

/** calculateTotal()
 *
 * Calculate subtotal plus tax, plus shipping cost if provided.
 * @param {string} subtotal
 * @param {string} state
 * @param {string||number} shippingCost
 *
 * returns: string
 */

export function calculateTotal(subtotal, state, shippingCost) {
  subtotal = new Decimal(subtotal);
  const tax = calculateTax(subtotal, state);
  shippingCost = shippingCost || 0;
  return subtotal.plus(tax).plus(shippingCost).toFixed(2);
}
