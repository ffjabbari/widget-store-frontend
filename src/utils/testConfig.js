/** Test data setup/teardown helper hooks */

export const TEST_DATA = {};

export const populateTestDataHook = testData => {
  testData.product = {
    byline: 'Latest generation smart televison 2021',
    date_added: '2020-12-29T08:00:00.000Z',
    description:
      'Lorem ipsum dolor sit, amet consectetur adipisicing elit. Nihil, voluptates enim nam voluptatum, quia doloribus quis sequi, porro consequatur excepturi eos error voluptas nobis modi atque magni commodi amet placeat.',
    discount: '0.00',
    image_url:
      'https://www.lg.com/sg/images/tvs/md05271244/gallery/LG-43LH600T-L.jpg',
    name: 'Samsung XG-900 55 inch LCD TV',
    price: '400.40',
    product_id: 1,
    quantity: 100,
    rating: null,
  };

  testData.product2 = {
    byline: 'The only remote you will need',
    description: 'Lorem ipsum dolor sit, placeat.',
    discount: '0.15',
    image_url:
      'https://cnet3.cbsistatic.com/img/SB-Lc9lni02k-GQeFZlVGII6cpg=/1070x602/2010/08/03/57369690-bb77-11e2-8a8e-0291187978f3/33981226-2-1333-OVR-1.jpg',
    name: 'Logitech Singularity Universal Remote',
    price: '40.20',
    product_id: 2,
    quantity: 1,
    rating: null,
  };

  testData.userProfileData = {
    data: {
      user: {
        user_id: 1,
        username: 'widgetLord',
        email: 'foo@gmail.com',
        first_name: 'bob',
        last_name: 'bobert',
        address: '123 Hill Ave.',
        address_line2: null,
        city: 'Big City',
        state: 'Texas',
        postal_code: '33333',
        phone_number: null,
        avatar_url: null,
        orders: [
          {
            order_id: 1,
            order_date: '2021-01-24T08:00:00.000Z',
            status: 'Pending',
          },
        ],
      },
    },
  };

  testData.CustomerInfoFormsValues = {
    first_name: 'Test',
    last_name: 'User',
    email: 'foo@gmail.com',
    address: '123 Main St',
    address_line2: 'Apt 5',
    city: 'Big City',
    state: 'TX',
    postal_code: '39488',
    phone_number: '(555) 555-5555',
  };

  testData.cart = {
    items: {
      2: {
        byline: 'The only remote you will need',
        description: 'Lorem ipsum dolor sit, placeat.',
        discount: '0.15',
        image_url:
          'https://cnet3.cbsistatic.com/img/SB-Lc9lni02k-GQeFZlVGII6cpg=/1070x602/2010/08/03/57369690-bb77-11e2-8a8e-0291187978f3/33981226-2-1333-OVR-1.jpg',
        name: 'Logitech Singularity Universal Remote',
        price: '40.20',
        product_id: 2,
        quantity: 1,
        rating: null,
      },
    },
    numCartItems: 1,
    subtotal: '34.17',
  };

  // mock return from /order/:id
  testData.orderGet = {
    customer: 1,
    customer_info: {
      first_name: 'Test',
      last_name: 'User',
      email: 'foo@gmail.com',
      address: '123 Main St',
      address_line2: '',
      city: 'Big City',
      state: 'TX',
      postal_code: '39488',
      phone_number: '(555) 555-5555',
    },

    discount: null,
    items: [
      { 0: { order_id: 2, product_id: 6, quantity: 1 } },
      { 1: { order_id: 2, product_id: 8, quantity: 1 } },
    ],

    order_date: '2021-02-05T08:00:00.000Z',
    order_id: 2,
    shipping_address: null,
    shipping_cost: '12.00',
    shipping_method: {
      details: { cost: '12.00', name: 'USPS ground shipping (3-7 days)' },
      shipping_method: 'usps_ground',
    },

    status: 'Pending',
    subtotal: '784.16',
    tax: '66.65',
    total: '862.81',
    total_items_quantity: 2,
    tracking_number: null,
  };

  TEST_DATA.shippingMethods = {
    usps_ground: {
      cost: '12.00',
      name: 'USPS ground shipping (3-7 days)',
    },
    usps_priority: {
      cost: '16.00',
      name: 'USPS priority shipping (2-4 days)',
    },
    usps_next_day: {
      cost: '25.00',
      name: 'USPS next day delivery (1 day)',
    },
    ups_ground: {
      cost: '16.00',
      name: 'UPS ground shipping (3-6 days)',
    },
    ups_2nd_day: {
      cost: '20.00',
      name: 'UPS second day delivery (2 days)',
    },
    ups_overnight: {
      cost: '28.00',
      name: 'UPS overnight delivery (1 day)',
    },
  };
};
