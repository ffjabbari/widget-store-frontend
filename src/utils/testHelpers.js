import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { testStore } from '../redux/store/reduxStore';

export const renderWithStore = component =>
  render(
    <Provider store={testStore}>
      <MemoryRouter>{component}</MemoryRouter>
    </Provider>
  );
