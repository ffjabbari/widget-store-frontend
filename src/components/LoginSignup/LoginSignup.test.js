// local imports
import { fireEvent } from '@testing-library/react';
import { renderWithStore } from '../../utils/testHelpers';
import LoginSignup from './LoginSignup';

const mockUseHistoryGoBack = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({
    goBack: mockUseHistoryGoBack,
  }),
}));

afterEach(() => jest.resetAllMocks());

test('should render', () => {
  renderWithStore(<LoginSignup />);
});

test('LoginSignup snapshot', async () => {
  const { asFragment } = renderWithStore(<LoginSignup />);
  expect(asFragment()).toMatchSnapshot();
});

test('close button calls history goBack', () => {
  const { getAllByRole } = renderWithStore(<LoginSignup />);

  const cancelBtn = getAllByRole('button', { name: /cancel/i });
  fireEvent.click(cancelBtn[0]);

  expect(mockUseHistoryGoBack.mock.calls.length).toBe(1);
});
