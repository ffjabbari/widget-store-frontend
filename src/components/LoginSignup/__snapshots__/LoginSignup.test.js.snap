// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginSignup snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="rounded col-md col-sm-12 offset-md-3"
      >
        <div />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
